test/evp_extra_test.c

```diff
@@ -447,19 +548,38 @@ static int test_EVP_DigestSignInit(void)
             || !TEST_true(EVP_DigestSignFinal(md_ctx, sig, &sig_len)))
         goto out;
 
+    if (tst >= 2) {
+        if (!TEST_int_gt(BIO_reset(mdbio), 0)
+                || !TEST_int_gt(BIO_get_md_ctx(mdbio, &md_ctx_verify), 0))
+            goto out;
+    }
+
     /* Ensure that the signature round-trips. */
     if (!TEST_true(EVP_DigestVerifyInit(md_ctx_verify, NULL, EVP_sha256(),
-                                        NULL, pkey))
-            || !TEST_true(EVP_DigestVerifyUpdate(md_ctx_verify,
-                                                 kMsg, sizeof(kMsg)))
-            || !TEST_true(EVP_DigestVerifyFinal(md_ctx_verify, sig, sig_len)))
+                                        NULL, pkey)))
+        goto out;
+
+    if (tst >= 2) {
+        if (!BIO_write_ex(mdbio, kMsg, sizeof(kMsg), &written))
+            goto out;
+    } else {
+        if (!TEST_true(EVP_DigestVerifyUpdate(md_ctx_verify, kMsg,
+                                              sizeof(kMsg))))
+            goto out;
+    }
+    if (!TEST_true(EVP_DigestVerifyFinal(md_ctx_verify, sig, sig_len)))
         goto out;
 
     ret = 1;
 
  out:
-    EVP_MD_CTX_free(md_ctx);
-    EVP_MD_CTX_free(md_ctx_verify);
+    if (tst >= 2) {
+        BIO_free(membio);
+        BIO_free(mdbio);
+    } else {
+        EVP_MD_CTX_free(md_ctx);
+        EVP_MD_CTX_free(md_ctx_verify);
+    }
```

Comment:
Since all these variables are now properly initialized with NULL, there should be no more need to check `tst` here.

Reply:
Err, no. We still need the check. Depending on which test we are running we have either created the ctx values ourselves or got them out of the BIO. If they are out of the BIO then we are not the owner so we should not free them.

Reply:
Ah. So the trade-off is to have a couple of extra variables to hold the pointers this function owns, and any time you do allocate a context, you do something like this:

``` C
    if (!TEST_ptr(md_ctx = a_md_ctx = EVP_MD_CTX_new())
```

... and then just free `a_md_ctx`

The benefit that method is avoiding having to remember what condition what variable was given what value ; it's less error prone.

Reply:
Yes, it can certainly be done that way. Fixup commit pushed.